src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs

```diff
@@ -179,6 +180,11 @@ public void RegisterWorkspaceHost(IVisualStudioWorkspaceHost host)
             _workspaceHosts.Add(new WorkspaceHostState(this, host));
         }
 
+        public Task RegisterWorkspaceHostFromForegroundThreadAsync(IVisualStudioWorkspaceHost host)
```

Comment:
it is ProjectTracker's requirement for RegisterWorkspaceHost to be called from UI thread. not sure why all caller need to deal with it when ProjectTracker can make sure that happens itself.

alternative approach is making RegisterWorkspaceHost to RegisterWorkspaceHostAsync but having sync and async version seems better since two just can't be mix used.